Kingart curtain switch UI support

echauvet opened the pull request:
This is a preliminary commit to extend KA work by @AlbertWeterings. Using AT+START commands, we know the real position of the curtain. Small MCU bugs are bypassed as commented in the code.
The UI is animated  and displays :
- A basic animation of the curtain opening (4 types available) - this reflects the real position of the curtain if the switch is correctly calibrated ( the MCU adds 2 seconds to the initial calibration)
- The 3 buttons are displayed and animated (blue/red colors) as the curtain switch. (this is achieved with AT+START commands and move status detection)
- A slider to set an absolute position. Its position is refreshed whenever the curtain stops moving.

![image](https://user-images.githubusercontent.com/4821565/81844992-378aff80-9550-11ea-9367-db68ce49aa1c.png)

A configuration panel as been created to choose:
- The curtain type. I will use it later for mqtt to integrate a basic UI layout of the curtain
![image](https://user-images.githubusercontent.com/4821565/81845434-d57eca00-9550-11ea-8a83-267b4f5e33b7.png)
- Once rebooted/powered on, the curtain initial position could be set
![image](https://user-images.githubusercontent.com/4821565/81845477-e62f4000-9550-11ea-8017-da6916f7b126.png)

Next steps are code cleaning, integrate any reviews from contributors and improve the MQTT stack with  RedNode. For now, I can drive my curtain as a Google curtain, or a switch or a lamp bulb (for brightness setting). I'm preparing some demo flow with Google legacy curtain, and the hybrid solution.

Both core base and latest-base are working - with latest-base, a two steps upgrade is required.
I'm cheking the work done for OTA in RedNode to check if some code is ready for 2 steps upgrade.

All Tests has been done using a curtain switch, espurna UI, google home and RedNode UI and all seems to animate together without bugs.

echauvet committed:
Completed kingart curtain switch support:
Added Web UI controls and view in status
Added power up behaviour (nothing, close, open, last position)
Added curtain style (Roller, etc...) for UI status and to be used in MQTT

echauvet committed:
Checked commit before PR. Added html gulp static files generated.

echauvet committed:
Corrected a code refactor error which make the MQTT messages not handling position from received messages. Added comment to debug tab to show how to enable debug

echauvet committed:
Use the maximum compression for gzip. A byte is a byte :)

mcspr commented on code/html/custom.js:
```diff
@@ -1691,6 +1734,49 @@ function processData(data) {
+            }
+            $("#curtainSet").val(value.curtainSet); //Update sliders
+            if(value.curtainMoving == "Idle") { //When idle, all buttons are off (blue)
+                $("button.curtain-button").css('background', 'rgb(66, 184, 221)');
```
Consider adding a vars for both rgb(...) as value explanation instead of comment?

Similar for background, inline func definition with an arg, perhaps?
```js
var gradient = function(degrees, pair1, pair2) { $("#curtainGetPicture").css('background', ...); };
// for case '3', just check if `pair2 !== undefined`
```

echauvet replied:
It seems to increase the footprint, no?

mcspr replied:
The idea was to shorten the string where the action happens. We can easily see that values are actually set

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -24,21 +29,157 @@ Copyright (C) 2019 by Albert Weterings
+bool _curtain_moving                = true;
+
+//Enable more traces, true as a default and stopped when curtain is setup.
+bool _debug_flag = true;
```
`_curtain` prefix?

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -24,21 +29,157 @@ Copyright (C) 2019 by Albert Weterings
+void _KACurtainSet(int button, int position = CURTAIN_POSITION_UNKNOWN) {
+
+   if(_curtain_waiting_ack) {
+        DEBUG_MSG_P(PSTR("[KAUART] MCU BUSY : Request ignored!\n"));
```
I noticed that we not always respect debug_flag.
Maybe add a file-level function / macro at the top that wraps DEBUG_MSG_P with debug_flag? e.g.
```cpp
#define KINGART_DEBUG_MSG_P(...) do { if (_curtain_debug_flag) { DEBUG_MSG_P(__VA_ARGS__); } } while(0)
```

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+    //Update last position and transmit to MQTT (GUI is at the end)
+    if(_curtain_position != CURTAIN_POSITION_UNKNOWN && _curtain_last_position != _curtain_position) {
+        _curtain_last_position = _curtain_position;
+        if(_curtain_initial_position_set) //Update initial position - TDOD : maybe only when move is finished to avoid to write too frequently
```
re TODO, you can always use rtcmem to store intermediate value and commit when it settles
see lights for example, where ticker is reset each time the value changes. setsetting only happens when values don't change for a while

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+//------------------------------------------------------------------------------
+void _curtainWebSocketUpdate(JsonObject& root) {
+    JsonObject& state = root.createNestedObject("curtainState");
+    state["curtainGet"] = _curtain_last_position;
```
No need to prefix with `curtain` when we already know we are in `curtainState`?

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -24,21 +29,157 @@ Copyright (C) 2019 by Albert Weterings
+             _KACurtainSet(CURTAIN_BUTTON_CLOSE);
+        else if (init_position == 2) 
+             _KACurtainSet(CURTAIN_BUTTON_OPEN);
+        else if (init_position == 3) {
```
I think we are missing explanation of curtain behaviour numbers?

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+        _curtain_position_set = _curtain_last_position;
+    state["curtainSet"] = _curtain_position_set;
+    state["curtainButton"] = _curtain_last_button;
+    state["curtainMoving"] = _curtain_moving ? "Moving" : "Idle";
```
`bool` directly? there are only two possible 'states'
same for js, no need to compare string.

echauvet replied:
Done

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+                _debug_flag = true;
+            } else
+                _KACurtainSend(data["command"]);
+    }
```
This should be handled by terminal commands?
`terminalRegisterCommand(F("KAUART.DEBUG"), ...)`
`terminalRegisterCommand(F("KAUART.SEND"), ...)`?

echauvet replied:
I need more time to understand this.

echauvet replied:
Let see this later - Terminal command for this device?

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -24,21 +29,157 @@ Copyright (C) 2019 by Albert Weterings
+    else if( _curtain_last_position != CURTAIN_POSITION_UNKNOWN)
+        _curtain_position_set = _curtain_last_position;
+
+    if(_curtain_initial_position_set == false) {//The curtain stopped moving for the first time, set the position back to 
```
```suggestion
    if (!_curtain_initial_position_set) { //The curtain stopped moving for the first time, set the position back to 
```

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+
+//------------------------------------------------------------------------------
+void _curtainWebSocketStatus(JsonObject& root) {
+    root["curtainVisible"] = 1;
```
Already set in onvisible

echauvet replied:
Right

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,108 +235,248 @@ To configure the device:
+//------------------------------------------------------------------------------
+void _curtainWebSocketOnConnected(JsonObject& root) {
+    root["curtainType"] = getSetting("curtainType", "0");
+    root["curtainInitialBehaviour"] = getSetting("curtainInitialBehaviour", "0");
```
`curtainInitialBehaviour` -> `curtainBoot`?
"0" is only explained in web

echauvet replied:
curtainBt/curtainBtPos

mcspr reviewed (changes requested):
Thanks!

- If you don't mind going over code comments once more, please verify that those still apply.
- We should explain magic numbers in code (and, if possible, in log too). I've added some helpers for getSetting() 2nd arg, we can map any type from string. see `namespace settings { namespace internal` in relays / buttons / sensors.
- We don't enforce any specific style / formatting... But one thing I am not a fan of is `if (cond) expr;`, unless those are one-liners like `if (something) return;`. Having brackets is preferable.

echauvet committed:
Update code/espurna/curtain_kingart.cpp

Co-authored-by: Max Prokhorov <prokhorov.max@outlook.com>

echauvet committed:
Variable name prefix update
Added public set position function and curtain count

echauvet committed:
Added Scheduler new type : curtain

echauvet committed:
Add curtain schedule support and correct the 12h bug when adding a slot (00:00 after refresh). Default Hour is 12h

echauvet committed:
Code cleaning.
Schedule support.
Light support is having an issue with new schedules corrected here.

echauvet committed:
I let do better by the specialists

echauvet committed:
Curtain init statues

mcspr commented on code/espurna/scheduler.cpp:
```diff
@@ -62,7 +71,7 @@ void _schWebSocketOnConnected(JsonObject &root){
         action.add(getSetting({"schAction", i}, 0));
         type.add(getSetting({"schType", i}, SCHEDULER_TYPE_SWITCH));
-        hour.add(getSetting({"schHour", i}, 0));
+        hour.add(getSetting({"schHour", i}, 12));
```
Can you explain why change to 12?

echauvet replied:
The default value in the custom.js UI is set to 12h, if not edited, the final stored value will be 0h even if you use save button. After reboot it will be 0h.
I added schType in change listening too. It seems that light schedule suffer from a bug which stores a light schedule as switch schedule.

mcspr replied:
Instead of changing default, we should change UI to display the correct value.
This looks like bug from #1986, 12 should not have been there

mcspr commented on code/espurna/scheduler.cpp:
```diff
@@ -102,15 +111,15 @@ void _schConfigure() {
                 bool sch_enabled = getSetting({"schEnabled", i}, false);
                 int sch_action = getSetting({"schAction", i}, 0);
-                int sch_hour = getSetting({"schHour", i}, 0);
+                int sch_hour = getSetting({"schHour", i}, 12);
```
Can you explain why change to 12?

echauvet replied:
Same as previous answer

mcspr commented on code/espurna/scheduler.cpp:
```diff
@@ -230,7 +246,7 @@ void _schCheck(int relay, int daybefore) {
         if (_schIsThisWeekday(sch_utc ? calendar_weekday.utc_wday : calendar_weekday.local_wday, sch_weekdays)) {
 
-            int sch_hour = getSetting({"schHour", i}, 0);
+            int sch_hour = getSetting({"schHour", i}, 12);
```
Can you explain why change to 12?

echauvet replied:
Same as previous answer

mcspr commented on code/espurna/curtain_kingart.cpp:
```diff
@@ -87,112 +243,266 @@ To configure the device:
+                _curtain_debug_flag = true;
+            } else
+                _KACurtainSend(data["command"]);
+    }
```
re Terminal, you **only** handle WebUI this way, while command handles every input from Telnet / http / webui / etc.
(and which also tries to handle the command that you send it, you just don't see the error message there)

Commands are akin to the main() func, we get argc and argv[],so argv[0] == command name and argc > 1 means that we have some params.

echauvet replied:
Removed it from now

mcspr commented on code/espurna/scheduler.cpp:
```diff
@@ -311,7 +337,7 @@ void schSetup() {
         static bool restore_once = true;
         if (restore_once) {
-            for (unsigned char i = 0; i < relayCount(); i++) {
+            for (unsigned char i = 0; i < schedulableCount(); i++) {
```
Doesn't this step on schCheck id resolution? If we have 1 relay and 1 curtain, for example.

echauvet replied:
I've done a big answer and as a result:
![image](https://user-images.githubusercontent.com/4821565/82129570-37933580-97c4-11ea-9bea-3b8cf974495c.png)
 So the quick answer is that I tested all with a dummy switch in hardware.h and you can choose between both types and it works:
![image](https://user-images.githubusercontent.com/4821565/82129613-75905980-97c4-11ea-9a22-11da094ed893.png)

Switch and curtain schedule task worked.
This is how the light schedule bug has been detected (it stays as a switch schedule instead of light schedule)

mcspr replied:
In this specific block, schCheck will recursively check every 'rule' that uses specific relayID (index from old relayCount()) and goes back a week day-by-day to check if rule applies in the past.
Although, we don't count light channels, so it's still sort-of broken because of non-matching ID / indexes.

echauvet replied:
Understood! In fact, we need specific processing for curtain.

mcspr reviewed (commented):
Let's think a bit more about the scheduler stuff? I am not sure restoring works like that, but I'd expect normal operation mode does. This can be a separate patch, not to complicate things further.

If you are ok with kingart .cpp stuff, I'll reformat it for the merge.

echauvet commented:
<!--
/* Font Definitions */
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
/* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
a:link, span.MsoHyperlink
	{mso-style-priority:99;
	color:blue;
	text-decoration:underline;}
pre
	{mso-style-priority:99;
	mso-style-link:"Pr\00E9format\00E9 HTML Car";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:10.0pt;
	font-family:"Courier New";}
span.PrformatHTMLCar
	{mso-style-name:"Pr\00E9format\00E9 HTML Car";
	mso-style-priority:99;
	mso-style-link:"Pr\00E9format\00E9 HTML";
	font-family:"Courier New";}
.MsoChpDefault
	{mso-style-type:export-only;}
@page WordSection1
	{size:612.0pt 792.0pt;
	margin:70.85pt 70.85pt 70.85pt 70.85pt;}
div.WordSection1
	{page:WordSection1;}
-->How do we do ? I start a correction, you start it ? let me know.   De : Max ProkhorovEnvoyé le :samedi 16 mai 2020 22:14À : xoseperez/espurnaCc : Eric Chauvet; AuthorObjet :Re: [xoseperez/espurna] Kingart curtain switch UI support (#2250) @mcspr commented on this pull request.In code/espurna/scheduler.cpp:> @@ -62,7 +71,7 @@ void _schWebSocketOnConnected(JsonObject &root){         switch_.add(getSetting({"schSwitch", i}, 0));         action.add(getSetting({"schAction", i}, 0));         type.add(getSetting({"schType", i}, SCHEDULER_TYPE_SWITCH));-        hour.add(getSetting({"schHour", i}, 0));+        hour.add(getSetting({"schHour", i}, 12));Instead of changing default, we should change UI to display the correct value.This looks like bug from #1986, 12 should not have been there—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe. 

echauvet committed:
Coding style corrections

echauvet committed:
Removed debug stuff from now - will be back later with the right way

AlbertWeterings commented:
Great work you are doing Eric, your work makes the Curtain switch usable for many people that are not using it only over MQTT. I did not have the knowledge and time to build the web interface so I took off after MQTT was finished and my 9 shutters where moving the way I wanted them to move.

In your code I fount this:
    // Need to send confirmation to the N76E003AT20 that message is received
    // ECH : TODO Check this is the case every time
    _KACurtainSend("AT+SEND=ok");

I can confirm it is mandatory to confirm to the N76E003AT20 that every message send by it is received. When I started working on it I found that after power up the switch it takes about 30sec before the N76E003AT20 starts sending messages (just not log enough to flash the ESP) so at first I had jumpers on the data lines to disconnect both chips from each other and be able to flash ESPURNA. After silencing the N76E003AT20 flashing is possible. Now a day's we have to OTA flash in two stages this is possible as long the N76E003AT20 stay's quiet after a power cycle flashing is not possible anymore as the N76E003AT20 is corrupting the flash process.

I have been logging communication between both chips with the original software for quite some time to fully understand how it works.

mcspr commented:
> Now a day's we have to OTA flash in two stages this is possible as long the N76E003AT20 stay's quiet after a power cycle flashing is not possible anymore as the N76E003AT20 is corrupting the flash process.

re. some discussions in the Core repository and general observations about reboot.
Perhaps we need to do Serial.end() before doing OTA? Chip persist some of the configuration between sw reboots, including pins state / uart / etc.

AlbertWeterings commented:
b.t.w. Flashing over serial also fails as the N76E003AT20 starts sending data during flashing. This can be solved by holding it's pin 4 or on the rear side of the board RES1 to GND this keeps the N76E003AT20 in reset state.

mcspr committed:
rebuild webui

mcspr committed:
revert 435f1c5e03ac8ff3e5591795e67c00aef55dc956 schHour default, rebuild ui

mcspr reviewed (approved):
Merging as-is b/c curtain part is done, delaying scheduler changes (?) for later.

echauvet commented:
@AlbertWeterings the solution I found is to calibrate the curtain to a very long period for intial flash so the MCU does not send messages until the closed position is not reached. Do you confirm that this is the AT+UPDATE message "setclose":0 "switch":"off" sent at the end of the first move which breaks the update? If so, just calibrate curtain opening to > 60s - this will be enough to keep MCU quiet.

AlbertWeterings commented:
@echauvet This is a nice work around at the moment you are blocked from flashing after something went wrong. I think good to document. I preferer to silence the MCU in the software by confirming the last message is understood. I did a quick test, you can preform a OTA 2 stage update without problems when ESP is rebooting the MCU doesn't. Only if someone is touching the buttons the MCU starts repeating it's last message and flashing is not possible. Disabling the serial port during OTA flash procedure as @mcspr referred to is the best sollution.
